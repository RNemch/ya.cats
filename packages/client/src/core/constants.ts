export const routerPaths = {
  login: '/login',
  signup: '/signup',
  profile: '/profile',
  game: '/game',
  main: '/',
  leaderBoard: '/leaderboard',
  forum: '/forum',
  forumTopic: '/forum/:topicId',
}

export const defaultError = 'Something went wrong...Please try again'
