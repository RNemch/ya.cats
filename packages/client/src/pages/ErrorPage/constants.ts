import { ErrorPageType, ErrorContent } from './types'

export const errorContent: Record<ErrorPageType, ErrorContent> = {
  '500': {
    title: '500',
    description: 'INTERNAL SERVER ERROR',
  },
  '404': {
    title: '404',
    description: 'PAGE NOT FOUND',
  },
  errorBoundary: {
    title: 'Something went wrong...',
    description: 'Please reload the page',
  },
}
