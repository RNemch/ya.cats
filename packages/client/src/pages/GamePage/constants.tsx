export const GAME_BACKGROUNDS = [
  'background_layer_3.png',
  'background_layer_2.png',
  'background_layer_1.png',
]
