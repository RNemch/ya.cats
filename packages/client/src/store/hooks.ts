import {
  TypedUseSelectorHook,
  useDispatch,
  useSelector,
  useStore,
} from 'react-redux'
import store, { AppDispatch, RootState } from '.'

export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
export const useAppStore: () => typeof store = useStore
