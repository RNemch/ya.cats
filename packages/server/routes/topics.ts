import express from 'express'
const router = express.Router()

const topics = [
  {
    id: 1,
    topicName: 'Doe',
    comments: [{ id: 1, username: 'john', text: 'new topic name' }],
  },
  {
    id: 2,
    topicName: 'Smith',
    comments: [{ id: 1, username: 'carol', text: 'text for topic' }],
  },
]

router.get('/', (_, res) => {
  res.send(topics)
})

router.get('/:id', (req, res) => {
  const { id } = req.params
  res.send(topics.find(topic => topic.id === +id))
})

router.post('/add', (req, res) => {
  const { topicName } = req.body

  topics.push({ id: topics.length + 1, topicName, comments: [] })

  res.send(topics)
})

router.post('/comment/add', (req, res) => {
  const { topicId, username, text } = req.body

  const id = topics.findIndex(topic => topic.id === +topicId)
  const topicToUpdate = {
    ...topics[id],
    comments: [
      ...topics[id].comments,
      {
        id: topics[id].comments.length + 1,
        username: username,
        text: text,
      },
    ],
  }

  topics.splice(topicId, 1, topicToUpdate)

  res.send(topics)
})

export default router
