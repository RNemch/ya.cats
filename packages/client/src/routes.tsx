import ErrorBoundary from '@components/error-boundary'
import ProtectedRoute from '@components/protected-route'
import { routerPaths } from '@core/constants'
import ErrorPage from '@pages/ErrorPage'
import ForumPage, { initForumPage } from '@pages/ForumPage'
import ForumTopicPage, { initForumTopicPage } from '@pages/ForumTopicPage'
import GamePage, { initGamePage } from '@pages/GamePage'
import LeaderBoardPage, { initLeaderBoardPage } from '@pages/LeaderBoardPage'
import LoginPage from '@pages/LoginPage'
import MainPage, { initMainPage } from '@pages/MainPage'
import ProfilePage, { initProfilePage } from '@pages/ProfilePage'
import SignupPage from '@pages/SignupPage'
import { AppDispatch, RootState } from './store'

export type PageInitContext = {
  authCookie?: string
  uuid?: string
}

export type PageInitArgs = {
  dispatch: AppDispatch
  state: RootState
  ctx: PageInitContext
}

export type InitRoutes = {
  path: string
  fetchData: ({ dispatch, state, ctx }: PageInitArgs) => Promise<unknown>
}

export const routes = [
  {
    path: '/',
    element: (
      <ErrorBoundary>
        <ProtectedRoute authProtected={false} />
      </ErrorBoundary>
    ),
    children: [
      {
        path: routerPaths.login,
        element: <LoginPage />,
      },
      {
        path: routerPaths.signup,
        element: <SignupPage />,
      },
    ],
  },
  {
    path: '/',
    element: (
      <ErrorBoundary>
        <ProtectedRoute authProtected={true} />
      </ErrorBoundary>
    ),
    children: [
      {
        index: true,
        element: <MainPage />,
      },
      {
        path: routerPaths.profile,
        element: <ProfilePage />,
      },
      {
        path: routerPaths.game,
        element: <GamePage />,
      },
      {
        path: routerPaths.leaderBoard,
        element: <LeaderBoardPage />,
      },
      {
        path: routerPaths.forum,
        element: <ForumPage />,
      },
      {
        path: routerPaths.forumTopic,
        element: <ForumTopicPage />,
      },
    ],
  },
  {
    path: '*',
    element: <ErrorPage type="404" />,
  },
]

export const initRoutes: InitRoutes[] = [
  {
    path: routerPaths.main,
    fetchData: initMainPage,
  },
  {
    path: routerPaths.profile,
    fetchData: initProfilePage,
  },
  {
    path: routerPaths.game,
    fetchData: initGamePage,
  },
  {
    path: routerPaths.leaderBoard,
    fetchData: initLeaderBoardPage,
  },
  {
    path: routerPaths.forum,
    fetchData: initForumPage,
  },
  {
    path: routerPaths.forumTopic,
    fetchData: initForumTopicPage,
  },
]
