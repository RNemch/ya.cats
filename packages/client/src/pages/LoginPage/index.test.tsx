import { Provider } from 'react-redux'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'

import { routes } from '../../routes'
import { routerPaths } from '@core/constants'
import store from '@store/index'

describe('Login Page', () => {
  let loginInput: HTMLInputElement
  let passwordInput: HTMLInputElement
  let submitButton: HTMLButtonElement

  const router = createMemoryRouter(routes, {
    initialEntries: [routerPaths.login],
  })

  const wrapper = (
    <Provider store={store}>
      <RouterProvider router={router} />
    </Provider>
  )

  test('Должен верно отрисоваться', () => {
    const { container } = render(wrapper)

    expect(container).toMatchSnapshot()
  })

  describe('Проверка функциональности', () => {
    beforeEach(() => {
      render(wrapper)

      loginInput = screen.getByLabelText('Login')
      passwordInput = screen.getByLabelText('Password')
      submitButton = screen.getByText('LOG IN')
    })

    test('Должны быть показаны ошибки при пустых полях', async () => {
      fireEvent.click(submitButton)

      await waitFor(() => {
        expect(screen.getByText('Login is required')).toBeDefined()
        expect(screen.getByText('Password is required')).toBeDefined()
      })
    })

    test('Должны быть показаны ошибки при неправильно заполненных полях', async () => {
      fireEvent.change(loginInput, { target: { value: 'aa' } })
      fireEvent.change(passwordInput, { target: { value: 'qwerty12' } })
      fireEvent.click(submitButton)

      await waitFor(() => {
        expect(screen.getByText('Incorrect login')).toBeDefined()
        expect(screen.getByText('Incorrect password')).toBeDefined()
      })
    })
  })
})
