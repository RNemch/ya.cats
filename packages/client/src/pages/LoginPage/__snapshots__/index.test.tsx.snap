// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Должен верно отрисоваться 1`] = `
<div>
  <div
    class="background "
  >
    <div
      class="center "
    >
      <div
        class="container"
      >
        <svg
          class="spinner"
          viewBox="0 0 50 50"
        >
          <circle
            class="path"
            cx="25"
            cy="25"
            fill="none"
            r="20"
            stroke-width="5"
          />
        </svg>
      </div>
      <div
        class="paper "
      >
        <div
          class="space "
          style="flex-direction: column; gap: 62px; align-items: center;"
        >
          <h2
            class="xxxl grey-with-shadow "
            style="text-align: center;"
          >
            LOGIN
          </h2>
          <form>
            <div
              class="space "
              style="flex-direction: column; gap: 62px; align-items: center;"
            >
              <div
                class="space "
                style="flex-direction: column; gap: 16px;"
              >
                <label
                  class="label  m"
                  style="width: 300px;"
                >
                  Login
                  <input
                    class="input"
                    name="login"
                    placeholder="Login"
                    style="height: 48px;"
                    type="text"
                  />
                  <span
                    class="error"
                  />
                </label>
                <label
                  class="label  m"
                  style="width: 300px;"
                >
                  Password
                  <input
                    class="input"
                    name="password"
                    placeholder="Password"
                    style="height: 48px;"
                    type="password"
                  />
                  <span
                    class="error"
                  />
                </label>
              </div>
              <div
                class="space "
                style="flex-direction: column;"
              >
                <div
                  class="space "
                  style="flex-direction: row; gap: 12px;"
                >
                  <button
                    class="button  orange l"
                    color="orange"
                    style="height: 41px; width: 240px;"
                    type="submit"
                    w="240px"
                  >
                    LOG IN
                  </button>
                  <button
                    class="button  orange l"
                    color="orange"
                    style="height: 41px; width: 41px;"
                    type="button"
                    w="41px"
                  >
                    <img
                      alt="log in with yandex"
                      class="oAuthButton"
                      src="/ya-icon.png"
                    />
                  </button>
                </div>
              </div>
            </div>
          </form>
          <p
            class="m grey "
            style="text-align: center;"
          >
            Don’t have an account yet?
            <a
              class="link"
              href="/signup"
            >
              Register
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
